### ISSUE TITLE
Does not work on partitioned tables (PG10+)

### ISSUE THREAD

**Issue by justinperkins** [activerecord-hackery/ransack#992]
Ransack 1.8.9
Rails 5.1.3
Main database: MySQL
Secondary database: Postgres 10.x

I was running into "table does not exist" issues on an older version of Ransack (1.8.2), saw the connection bug fix and upgraded to 1.8.9. I'm still receiving the same error, but through testing I can see that ActiveRecord::ConnectionAdapters::SchemaCache#data_source_exists? is saying the table does not exist, even though ActiveRecord::ConnectionAdapters::PostgreSQLAdapter does list the table in question. See below

```
> PartitionModel.connection.schema_cache.class
=> ActiveRecord::ConnectionAdapters::SchemaCache
> PartitionModel.connection.tables
=> ["ar_internal_metadata", "activities_m11y2018", "schema_migrations", "activities", "activities_m04y2018", "activities_m10y2018"]
> PartitionModel.connection.schema_cache.data_source_exists?(:activities)
=> false
> PartitionModel.connection.class
=> ActiveRecord::ConnectionAdapters::PostgreSQLAdapter
> PartitionModel.connection.schema_cache.data_source_exists?(:activities_m04y2018)
=> true
```

I suspect the answer might be "Ransack is not intended to be used on partitioned Postgres tables", but  I did not see any reported issues on partitions so wanted to at least log something.

Thanks for looking

**Comment by seanfcarroll** [comment]
It would be a great feature to have... having a table partioned across 2 database types is a bit extreme (I didn't know you could even do that!).

It would be really helpful if it was possible to make a repo with the setup and some seed data. PR's are of course also welcome.

**Comment by justinperkins** [comment]
Sorry to be confusing, it's not a table partitioned across databases, we have two databases and the non-primary database has a partitioned table in it.

**Comment by seanfcarroll** [comment]
Just looking at this again. Rails does not natively support partioned tables, but in theory if ActiveRecord supports the queries against the partioned tables then Ransack should also.

Rails 6 does add multiple _database_ support, but this won't necessarily help with partitions within a database.

Would it be possible to write a failing test on a branch to work from?

**Comment by blakemcintyre** [comment]
I've been digging into this as I'm having the same issue on Postgres 11 using Rails 5.1. Ransack checks `data_source_exists?` on https://github.com/activerecord-hackery/ransack/blob/master/lib/ransack/adapters/active_record/context.rb#L25 which is essentially generated from ActiveRecord on https://github.com/rails/rails/blob/5-1-stable/activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb#L618. They key thing here is ` scope[:type] ||= "'r','v','m'" # (r)elation/table, (v)iew, (m)aterialized view`. This is missing `'p'` for partitioned tables which is found in Rails 5.2 and later https://github.com/rails/rails/blob/d9dfb4fd2a54bd1c50f7061d0f70250791857bf7/activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb#L750.
I ran the query generated from `data_source_sql` with `'p'` added in as an option for `c.relkind` and my partitioned tables were included in the results.

The quick(?) fix to this would be to update to Rails 5.2 otherwise Ransack would need to find a different way to get the column type.

**Comment by seanfcarroll** [comment]
I think this should be closed. Any comments on this?

**Comment by blakemcintyre** [comment]
@seanfcarroll if you know of an easy fix for this in ransack that would awesome but I have a feeling it probably isn't an easy fix and as more people upgrade their rails version this will not be relevant any more.

**Comment by seanfcarroll** [comment]
Hi @blakemcintyre Ransack needs a lot of work and it will need to come from the community. A lot of people use this gem, especially via ActiveAdmin.

I agreed to become a maintainer, but my role is management of contributions, not development of fixes. The contributions either need to come from the community or else we can accept donations via Open Collective and the project can then hire developers via Upwork or the like.

Cheers! 😄

**Comment by justinperkins** [comment]
I'm fine with you guys closing it if you're sure it's fixed in 5.2. Otherwise, it seems like there should at least be a feature request for Ransack to support partitioned tables.

**Comment by seanfcarroll** [comment]
@justinperkins is it fixed now under 5.2 and Ransack 2.4.0 ?

**Comment by justinperkins** [comment]
> @justinperkins is it fixed now under 5.2 and Ransack 2.4.0 ?

My legacy env isn't on 2.4.0, so I can't confirm the fix.

**Comment by seanfcarroll** [comment]
> My legacy env isn't on 2.4.0, so I can't confirm the fix.

Closing this issue as assuming it is fixed. It can be re-opened if still a problem.
